### DISCUSSION TITLE
Usage report of the efficient_conv_bn_eval feature

### DISCUSSION THREAD

**Discussion by youkaichao** [open-mmlab/mmengine#1252] [General]
Please report here how enabling `efficient_conv_bn_eval` feature helps your training. We expect the feature to seamlessly reduce memory footprint without affecting the performance. Currently it is an optional feature, but if the feedback is good enough, we can turn this feature into default :)

**Comment by JolyonWu** [comment]
I used the maskrcnn model as a baseline to test the difference of turning on this function. When the model first started training, it could indeed save 13% of the  memory. But as the training continued, I used nvidia-smi to check the  memory. I found that the  memory continued to increase from 7GB. It soared until it occupied 24GB, but the  memory usage in the training log record remained as low as it was at the beginning

**Reply by youkaichao** [reply]
Hi, numbers shown in `nvidia-smi` is known to be unreliable. Please refer to the documentation of pytorch https://pytorch.org/docs/stable/notes/cuda.html#memory-management for details. In your case, I suppose there are some memory cache showing up in `nvidia-smi`. The log record might be a more faithful number for memory footprint.

**Reply by JolyonWu** [reply]
Thanks for your prompt reply, I'll check again!

**Reply by youkaichao** [reply]
@buxuewushu1314 for example, please check the `memory` field at https://download.openmmlab.com/mmdetection/v2.0/faster_rcnn/faster_rcnn_r101_fpn_2x_coco/faster_rcnn_r101_fpn_2x_coco_20200504_210455.log.json .

**Comment by buxuewushu1314** [comment]
Train resnet18 on mmpretrain-v1.0.2 using '-- cfg options effective'_ Conv_ Bn_ After eval="[backbone]" ', there is almost no savings memory. Do I need to specify a version for mmretrain, or do I need any other settings?
![image](https://github.com/open-mmlab/mmengine/assets/63645215/1d4b3bc4-59fa-4189-bf9d-80a552072d5b)

![image](https://github.com/open-mmlab/mmengine/assets/63645215/a9ce1dee-3392-4633-8063-40601ecf2d5f)

**Reply by youkaichao** [reply]
Can you show your full training config? e.g. which task you are training on, and the config file you are using.

In general, pre-training uses BN in train mode, where `efficient_conv_bn_eval` cannot help. As the name suggests, it works when you train with BN in eval mode, which often occurs in mmdetection training.

**Reply by buxuewushu1314** [reply]
Thanks for your reply，yeah，it should be train mode，so does mmpretrain provide an interface for setting BN in eval mode, how should to do?

**Reply by youkaichao** [reply]
Generally we don't pretrain in eval mode. Pretraining requires train mode.

**Reply by buxuewushu1314** [reply]
Thank you again. classification model was still effective during fine-tuning by setting 'norm_eval', testing shows that it can save 33.8% memory,good job!

**Reply by youkaichao** [reply]
Glad it helps :)

**Comment by Jano-rm-rf** [comment]
> 再次感谢你。通过设置norm_eval，分类模型在微调时仍然有效，测试表明可以节省33.8%的内存，干得好！



> Thank you again. classification model was still effective during fine-tuning by setting 'norm_eval', testing shows that it can save 33.8% memory,good job!

Hello, I have set "norm_val", but there is an error：torch.fx.proxy.TraceError: symbolically traced variables cannot be used as inputs to control flow,how should to do?

**Reply by youkaichao** [reply]
Which config do you use to train your model?

**Reply by Jano-rm-rf** [reply]
I used the DINO config and changed the backbone network to efficientnet

**Reply by youkaichao** [reply]
Can you give a link to the config and the command you use?

**Reply by Jano-rm-rf** [reply]
[https://github.com/Jano-rm-rf/Test/blob/main/DINO-Test6-Config.py](url)
command:  python tools/train.py configs/dino/DINO-Test6.py --cfg-options efficient_conv_bn_eval="[backbone]"

**Reply by Jano-rm-rf** [reply]
![image](https://github.com/open-mmlab/mmengine/assets/77326693/8b0512bd-a3d5-4d81-9316-73a652c29272)
![image](https://github.com/open-mmlab/mmengine/assets/77326693/5a76155f-d907-42ee-a995-dc1db73757f5)
code running environment：torch2.0.1, mmdet3.2.0,mmengine0.9.0
Is version incompatibility cause the error？

**Reply by hvlgo** [reply]
This is not a problem caused by version incompatibility. This is due to the current implementation of  efficient_conv_ bn_eval feature is based on the symbolic trace in torch.fx. Due to the limited ability of symbolic trace, it is not possible to handle expressions that are only known at runtime or depend on parameters such as values or shapes of input. In the future, we will use torch.compile to implement this feature, and there will be no such issue at that time. If you want to learn more about symbolic trace, you can check this link https://pytorch.org/docs/stable/fx.html#limitations-of-symbolic-tracing or https://zhuanlan.zhihu.com/p/644590863.

**Reply by hvlgo** [reply]
The specific reason for this issue is that your config uses eficientnetv2, and its initialization specifies the conv type as Conv2dAdaptivePadding (https://github.com/open-mmlab/mmpretrain/blob/main/mmpretrain/models/backbones/efficientnet_v2.py#L186). The forward function of Conv2dAdaptivePadding contains condition judgment of pad_h calculated based on x.size() (https://github.com/open-mmlab/mmcv/blob/main/mmcv/cnn/bricks/conv2d_adaptive_padding.py#L58).

**Reply by Jano-rm-rf** [reply]
Thanks for your reply, I will modify conv type and test it

**Comment by zhangcan-hust** [comment]
Can this method still work if the batch_size is set to 1 during training?

**Reply by youkaichao** [reply]
As long as your BN is in eval mode during training, this method should work. Nevertheless, smaller batchsize may lead to less benefit in memory reduction.

**Reply by zhangcan-hust** [reply]
I understand, thanks for your reply!

**Comment by makecent** [comment]
Tested on the SlowOnly model (`torch.hub.load("facebookresearch/pytorchvideo", model='slow_r50', pretrained=True`). It reduced the memory from **4576Mb** to **2623Mb**:

![image](https://github.com/open-mmlab/mmengine/assets/42603768/133eb535-c6d0-4cc5-b5b0-357eb50d19e6)
![image](https://github.com/open-mmlab/mmengine/assets/42603768/dbfb4644-b02b-47a9-ab1a-546047dded93)

**Reply by youkaichao** [reply]
Glad it helps :)

**Comment by phoenixdna** [comment]
I have encoutered a question, please help.
I am using cuda11.8 and pytorch2.0.1; and the command line  is:
```command line
 ./tools/dist_train.sh bisai.py 1 --cfg-options efficient_conv_bn_eval="[backbone]"
```
my dist_train.sh
```shell
#!/usr/bin/env bash

CONFIG=$1
GPUS=$2
NNODES=${NNODES:-1}
NODE_RANK=${NODE_RANK:-0}
PORT=${PORT:-29568}
MASTER_ADDR=${MASTER_ADDR:-"127.0.0.1"}

PYTHONPATH="$(dirname $0)/..":$PYTHONPATH \
CUDA_VISIBLE_DEVICES=0,1 python -m torch.distributed.launch \
    --nnodes=$NNODES \
    --node_rank=$NODE_RANK \
    --master_addr=$MASTER_ADDR \
    --nproc_per_node=$GPUS \
    --master_port=$PORT \
    $(dirname "$0")/train.py \
    $CONFIG \
    --launcher pytorch ${@:3}

```

Then I have the following exception:

``` plain text
10/16 03:46:39 - mmengine - INFO - Enabling the "efficient_conv_bn_eval" feature for sub-modules: ['backbone']
Traceback (most recent call last):
  File "/gemini/code/mmdet_bisai/./tools/train.py", line 133, in <module>
    main()
  File "/gemini/code/mmdet_bisai/./tools/train.py", line 129, in main
    runner.train()
  File "/root/miniconda3/lib/python3.10/site-packages/mmengine/runner/runner.py", line 1762, in train
    turn_on_efficient_conv_bn_eval(ori_model, modules)
  File "/root/miniconda3/lib/python3.10/site-packages/mmengine/model/efficient_conv_bn_eval.py", line 158, in turn_on_efficient_conv_bn_eval
    turn_on_efficient_conv_bn_eval_for_single_model(module)
  File "/root/miniconda3/lib/python3.10/site-packages/mmengine/model/efficient_conv_bn_eval.py", line 147, in turn_on_efficient_conv_bn_eval_for_single_model
    fx_model: fx.GraphModule = fx.symbolic_trace(model)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 1109, in symbolic_trace
    graph = tracer.trace(root, concrete_args)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 778, in trace
    (self.create_arg(fn(*args)),),
  File "/gemini/code/mmdet_bisai/mmdet/models/backbones/swin.py", line 835, in forward
    x, hw_shape = self.patch_embed(x)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 756, in module_call_wrapper
    return self.call_module(mod, forward, args, kwargs)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 467, in call_module
    ret_val = forward(*args, **kwargs)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 749, in forward
    return _orig_module_call(mod, *args, **kwargs)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/gemini/code/mmdet_bisai/mmdet/models/layers/transformer/utils.py", line 302, in forward
    x = self.adap_padding(x)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 756, in module_call_wrapper
    return self.call_module(mod, forward, args, kwargs)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 467, in call_module
    ret_val = forward(*args, **kwargs)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/_symbolic_trace.py", line 749, in forward
    return _orig_module_call(mod, *args, **kwargs)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/gemini/code/mmdet_bisai/mmdet/models/layers/transformer/utils.py", line 176, in forward
    pad_h, pad_w = self.get_pad_shape(x.size()[-2:])
  File "/gemini/code/mmdet_bisai/mmdet/models/layers/transformer/utils.py", line 169, in get_pad_shape
    pad_h = max((output_h - 1) * stride_h +
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/proxy.py", line 413, in __bool__
    return self.tracer.to_bool(self)
  File "/root/miniconda3/lib/python3.10/site-packages/torch/fx/proxy.py", line 276, in to_bool
    raise TraceError('symbolically traced variables cannot be used as inputs to control flow')
torch.fx.proxy.TraceError: symbolically traced variables cannot be used as inputs to control flow
```
I checked a lot of material and ask GPT, but got no answer. Please do help and thx in advance.

**Reply by youkaichao** [reply]
> bisai.py

this model might contain certain conv that cannot be traced by `torch.fx` .

**Reply by youkaichao** [reply]
if your code supports `torch.compile` , you can use `torch >= 2.2` , and set 

```python
from torch._inductor import config as inductor_config
inductor_config.efficient_conv_bn_eval_fx_passes = True
```
